File: packages/sdk-ts/tests/threshold.test.ts
/**
 * Tests for FROST(Ed25519) threshold signing in the TypeScript SDK. Mirrors
 * the Python tests/test_threshold.py. Requires the optional peer dependency
 * @vouch-protocol-official/core-wasm; skipped entirely when it is not
 * installed (built locally via `core/wasm/build-npm.sh` and linked with
 * `npm install ../../core/wasm/pkg` for a local test run), the same graceful
 * skip the Python SDK uses when its native library is unavailable.
 */

import { createRequire } from 'node:module';

import {
  Signer,
  ThresholdError,
  ThresholdSigner,
  Verifier,
  groupPublicKeyMultikey,
  thresholdAggregate,
  thresholdCommit,
  thresholdGenerateKey,
  thresholdSignShare,
} from '../src';
import * as threshold from '../src/threshold';

let coreWasmAvailable = true;
try {
  createRequire(import.meta.url).resolve('@vouch-protocol-official/core-wasm');
} catch {
  coreWasmAvailable = false;
}

describe.skipIf(!coreWasmAvailable)('threshold: FROST(Ed25519) ceremony', () => {
  test('2-of-3 signs and verifies as plain Ed25519 (self-verified inside the core)', async () => {
    // aggregate() self-verifies inside the core before returning, so a
    // resolved, non-throwing call is itself the proof that the resulting
    // signature is a valid, standard Ed25519 signature over the message.
    const generated = await threshold.generateKey(2, 3);
    expect(generated.shares).toHaveLength(3);

    const [share0, share1] = generated.shares;
    const round1A = await threshold.commit(share0);
    const round1B = await threshold.commit(share1);
    const commitments: Record<string, string> = {
      [share0.identifier]: round1A.commitments,
      [share1.identifier]: round1B.commitments,
    };

    const message = new TextEncoder().encode('charge api.bank invoices/42');
    const sigShare0 = await threshold.signShare(message, share0, round1A.nonces, commitments);
    const sigShare1 = await threshold.signShare(message, share1, round1B.nonces, commitments);
    const shares: Record<string, string> = { [share0.identifier]: sigShare0, [share1.identifier]: sigShare1 };

    const signature = await threshold.aggregate(message, commitments, shares, generated.groupPublicKey);
    expect(signature).toHaveLength(64);
  });

  test('generateKey rejects a below-minimum threshold', async () => {
    await expect(threshold.generateKey(1, 3)).rejects.toBeInstanceOf(Error);
  });

  test('groupPublicKeyMultikey encodes a standard Ed25519 multikey', async () => {
    const generated = await threshold.generateKey(2, 3);
    const mk = groupPublicKeyMultikey(generated.groupPublicKey);
    expect(mk.startsWith('z6Mk')).toBe(true);
  });

  test('ThresholdSigner runs the full ceremony and plugs into Signer.fromBackend', async () => {
    const generated = await thresholdGenerateKey(2, 3);
    const signer = await ThresholdSigner.create(generated.shares.slice(0, 2), generated.groupPublicKey);

    const vouchSigner = Signer.fromBackend(
      'did:web:agent.example',
      groupPublicKeyMultikey(generated.groupPublicKey),
      signer.sign
    );
    const credential = await vouchSigner.sign({ action: 'read', target: 't', resource: 'https://x/y' });
    const result = await Verifier.verify(credential, groupPublicKeyMultikey(generated.groupPublicKey));
    expect(result.isValid).toBe(true);
  });

  test('ThresholdSigner requires at least 2 shares', async () => {
    const generated = await thresholdGenerateKey(2, 3);
    await expect(ThresholdSigner.create(generated.shares.slice(0, 1), generated.groupPublicKey))
      .rejects.toBeInstanceOf(ThresholdError);
  });

  test('different subsets of the same group verify against the same public key', async () => {
    const generated = await thresholdGenerateKey(3, 5);
    const message = new TextEncoder().encode('same message, different signer subset');

    async function signWith(indices: number[]): Promise<Uint8Array> {
      const chosen = indices.map((i) => generated.shares[i]);
      const noncesById: Record<string, string> = {};
      const commitments: Record<string, string> = {};
      for (const share of chosen) {
        const round1 = await thresholdCommit(share);
        commitments[share.identifier] = round1.commitments;
        noncesById[share.identifier] = round1.nonces;
      }
      const sharesOut: Record<string, string> = {};
      for (const share of chosen) {
        sharesOut[share.identifier] = await thresholdSignShare(
          message, share, noncesById[share.identifier], commitments);
      }
      return thresholdAggregate(message, commitments, sharesOut, generated.groupPublicKey);
    }

    const sigA = await signWith([0, 1, 2]);
    const sigB = await signWith([2, 3, 4]);
    expect(sigA).toHaveLength(64);
    expect(sigB).toHaveLength(64);
  });
});

File: go-sidecar/signer/threshold_test.go
//go:build frost

// Tests for FROST(Ed25519) threshold signing. Requires CGO_ENABLED=1 and the
// native library built via ../build-native.sh (run: go test -tags frost ./...
// with LD_LIBRARY_PATH pointing at ../lib).
package signer

import (
	"crypto/ed25519"
	"encoding/base64"
	"testing"
)

func generateCommitSignAggregate(t *testing.T, minSigners, maxSigners uint16, indices []int, message []byte) (*GenerateKeyResult, []byte) {
	t.Helper()
	generated, err := ThresholdGenerateKey(minSigners, maxSigners)
	if err != nil {
		t.Fatalf("ThresholdGenerateKey: %v", err)
	}
	if len(generated.Shares) != int(maxSigners) {
		t.Fatalf("expected %d shares, got %d", maxSigners, len(generated.Shares))
	}

	noncesByID := make(map[string]string)
	commitments := make(map[string]string)
	for _, i := range indices {
		share := generated.Shares[i]
		round1, err := ThresholdCommit(share)
		if err != nil {
			t.Fatalf("ThresholdCommit: %v", err)
		}
		commitments[share.Identifier] = round1.Commitments
		noncesByID[share.Identifier] = round1.Nonces
	}

	sharesOut := make(map[string]string)
	for _, i := range indices {
		share := generated.Shares[i]
		sigShare, err := ThresholdSignShare(message, share, noncesByID[share.Identifier], commitments)
		if err != nil {
			t.Fatalf("ThresholdSignShare: %v", err)
		}
		sharesOut[share.Identifier] = sigShare
	}

	signature, err := ThresholdAggregate(message, commitments, sharesOut, generated.GroupPublicKey)
	if err != nil {
		t.Fatalf("ThresholdAggregate: %v", err)
	}
	return generated, signature
}

func TestTwoOfThreeSignsAndVerifiesAsPlainEd25519(t *testing.T) {
	message := []byte("charge api.bank invoices/42")
	generated, signature := generateCommitSignAggregate(t, 2, 3, []int{0, 2}, message)

	verifyingKey, err := base64.StdEncoding.DecodeString(generated.GroupPublicKey.VerifyingKey)
	if err != nil {
		t.Fatalf("decode verifying key: %v", err)
	}
	if !ed25519.Verify(verifyingKey, message, signature) {
		t.Fatal("FROST aggregate signature must verify as a plain Ed25519 signature")
	}
}

func TestDifferentSubsetOfSameGroupAlsoVerifies(t *testing.T) {
	message := []byte("same message, different signer subset")
	generated, err := ThresholdGenerateKey(3, 5)
	if err != nil {
		t.Fatalf("ThresholdGenerateKey: %v", err)
	}

	signWith := func(indices []int) []byte {
		noncesByID := make(map[string]string)
		commitments := make(map[string]string)
		for _, i := range indices {
			share := generated.Shares[i]
			round1, err := ThresholdCommit(share)
			if err != nil {
				t.Fatalf("ThresholdCommit: %v", err)
			}
			commitments[share.Identifier] = round1.Commitments
			noncesByID[share.Identifier] = round1.Nonces
		}
		sharesOut := make(map[string]string)
		for _, i := range indices {
			share := generated.Shares[i]
			sigShare, err := ThresholdSignShare(message, share, noncesByID[share.Identifier], commitments)
			if err != nil {
				t.Fatalf("ThresholdSignShare: %v", err)
			}
			sharesOut[share.Identifier] = sigShare
		}
		sig, err := ThresholdAggregate(message, commitments, sharesOut, generated.GroupPublicKey)
		if err != nil {
			t.Fatalf("ThresholdAggregate: %v", err)
		}
		return sig
	}

	sigA := signWith([]int{0, 1, 2})
	sigB := signWith([]int{2, 3, 4})

	verifyingKey, err := base64.StdEncoding.DecodeString(generated.GroupPublicKey.VerifyingKey)
	if err != nil {
		t.Fatalf("decode verifying key: %v", err)
	}
	if !ed25519.Verify(verifyingKey, message, sigA) {
		t.Fatal("subset A signature must verify")
	}
	if !ed25519.Verify(verifyingKey, message, sigB) {
		t.Fatal("subset B signature must verify")
	}
}

func TestWrongMessageFailsVerification(t *testing.T) {
	message := []byte("original message")
	generated, signature := generateCommitSignAggregate(t, 2, 3, []int{0, 1}, message)

	verifyingKey, err := base64.StdEncoding.DecodeString(generated.GroupPublicKey.VerifyingKey)
	if err != nil {
		t.Fatalf("decode verifying key: %v", err)
	}
	if ed25519.Verify(verifyingKey, []byte("tampered message"), signature) {
		t.Fatal("a signature over one message must not verify a different message")
	}
}

func TestThresholdGenerateKeyRejectsBadThreshold(t *testing.T) {
	if _, err := ThresholdGenerateKey(1, 3); err == nil {
		t.Fatal("expected error: min_signers must be >= 2")
	}
	if _, err := ThresholdGenerateKey(4, 3); err == nil {
		t.Fatal("expected error: min_signers must be <= max_signers")
	}
}

func TestThresholdSignerNeedsAtLeastTwoShares(t *testing.T) {
	generated, err := ThresholdGenerateKey(2, 3)
	if err != nil {
		t.Fatalf("ThresholdGenerateKey: %v", err)
	}
	if _, err := NewThresholdSigner(generated.Shares[:1], generated.GroupPublicKey); err == nil {
		t.Fatal("expected error: ThresholdSigner needs at least 2 key shares")
	}
}

func TestThresholdSignerPlugsIntoBackendSigner(t *testing.T) {
	generated, err := ThresholdGenerateKey(2, 3)
	if err != nil {
		t.Fatalf("ThresholdGenerateKey: %v", err)
	}
	thresholdSigner, err := NewThresholdSigner(generated.Shares[:2], generated.GroupPublicKey)
	if err != nil {
		t.Fatalf("NewThresholdSigner: %v", err)
	}

	verifyingKey, err := base64.StdEncoding.DecodeString(generated.GroupPublicKey.VerifyingKey)
	if err != nil {
		t.Fatalf("decode verifying key: %v", err)
	}

	s, err := NewBackend("did:web:agent.example.com", ed25519.PublicKey(verifyingKey), thresholdSigner.Sign, 300)
	if err != nil {
		t.Fatalf("NewBackend: %v", err)
	}

	cred, err := s.Sign(SignOptions{Action: "read", Target: "t", Resource: "https://x/y"})
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}

	ok, err := VerifyDataIntegrityProof(cred, verifyingKey)
	if err != nil {
		t.Fatalf("VerifyDataIntegrityProof: %v", err)
	}
	if !ok {
		t.Fatal("credential signed by the threshold signer failed verification")
	}
}
